duckdb-latest/rollfun-duckdb-latest.R

```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env Rscript
+
+cat("# rollfun-duckdb-latest.R\n")
+
+source("./_helpers/helpers.R")
+
+suppressPackageStartupMessages({
+  library("DBI", lib.loc="./duckdb-latest/r-duckdb-latest", warn.conflicts=FALSE)
+  library("duckdb", lib.loc="./duckdb-latest/r-duckdb-latest", warn.conflicts=FALSE)
+})
+ver = packageVersion("duckdb")
+#git = "" # set up later on after connecting to db
+task = "rollfun"
+solution = "duckdb-latest"
+fun = "over"
+cache = TRUE
+
+data_name = Sys.getenv("SRC_DATANAME")
+src_grp = file.path("data", paste(data_name, "csv", sep="."))
+cat(sprintf("loading dataset %s\n", data_name))
+
+on_disk = as.numeric(strsplit(data_name, "_", fixed=TRUE)[[1L]][2L])>=1e9
+uses_NAs = as.numeric(strsplit(data_name, "_", fixed=TRUE)[[1L]][4L])>0
+if (on_disk) {
+  print("using disk memory-mapped data storage")
+  con = dbConnect(duckdb::duckdb(), dbdir=tempfile())
+} else {
+  print("using in-memory data storage")
+  con = dbConnect(duckdb::duckdb())
+}
+
+ncores = parallel::detectCores()
+invisible(dbExecute(con, sprintf("PRAGMA THREADS=%d", ncores)))
+invisible(dbExecute(con, "SET experimental_parallel_csv=true;"))
+git = dbGetQuery(con, "SELECT source_id FROM pragma_version()")[[1L]]
+
+# first create and ingest the table.
+invisible(dbExecute(con, "CREATE TABLE y(id1 INT, id2 INT, id3 INT, v1 FLOAT, v2 FLOAT, weights FLOAT)"))
+invisible(dbExecute(con, sprintf("COPY y FROM '%s' (AUTO_DETECT TRUE)", src_grp)))
+
+# no enums in table, also as of now no NULLs in data, so uses_NAs handling not needed
+invisible(dbExecute(con, "ALTER TABLE y RENAME TO x"))
+
+print(in_nr<-dbGetQuery(con, "SELECT count(*) AS cnt FROM x")$cnt)
+invisible(dbExecute(con, "DROP TABLE IF EXISTS ans"))
+
+# window size
+w = in_nr/1e3
+wsmall = in_nr/1e4
+wbig = in_nr/1e2
+
+task_init = proc.time()[["elapsed"]]
+cat("rolling...\n")
+
+question = "mean" # q1
+sql = sprintf("CREATE TABLE ans AS SELECT avg(v1) OVER (ORDER BY id1 ROWS BETWEEN %d PRECEDING AND CURRENT ROW) AS v1 FROM x", w-1L)
```

Comment:
The same queries can be used for DataFusion or spark as well

Reply:
that's great. unfortunately my volunteering availability is limited at the moment, and adding even one similar solution is quite a bit of work... setting up its environment, modifying script, validating results against another tool, checking it passes _report/report.R checks. If all goes smooth it's couple hours of work, but I don't think it ever went smooth :)

Reply:
No worries, just stating that the syntax should be the same for those engines 👍